v3/search.go

```diff
@@ -243,25 +244,76 @@ func (s *search) optimizeGroup(loc memoLoc, required *physicalProps, parent *sea
 func (s *search) optimizeGroupExpr(loc memoLoc, required *physicalProps, parent *searchTask) {
 	g := &s.memo.groups[loc.group]
 	e := &g.exprs[loc.expr]
-	if e.physicalProps == 0 {
-		// A logical expression, no need to optimize.
+	if (e.info().kind() & physicalKind) == 0 {
+		// We can only perform optimization on physical operators.
 		return
 	}
+
 	if !s.memo.physicalProps[e.physicalProps].provides(required) {
-		// TODO(peter): the enforcer mechanism here needs to be generalized.
+		// TODO(peter): The enforcer mechanism here needs to be generalized.
+		//
+		// TODO(peter): This isn't correct. We can only enforce an ordering on
+		// columns that are provided by the expression.
 		sort := &expr{
-			op:    sortOp,
-			loc:   memoLoc{loc.group, -1},
+			op:  sortOp,
+			loc: memoLoc{loc.group, -1},
+			children: []*expr{
+				&expr{
+					loc: loc,
+				},
+			},
 			props: g.props,
 			physicalProps: &physicalProps{
 				providedOrdering: required.providedOrdering,
 			},
-			private: &sortSpec{
-				loc: loc,
-			},
 		}
 		s.memo.addExpr(sort)
+		s.optimizeGroupTask(g, required, parent)
+
+		// TODO(peter): While the expression did not provide the required
```

Comment:
I don't understand this comment. Can you give an example?

Reply:
`SELECT * FROM a WHERE y > 1 ORDER by x`. The `WHERE y > 1` clause does not provide the desired ordering calling us to fall into this block and create a "sort" enforcer. But we also asked the input (`SELECT * FROM a`) for the best plan that provides the desired ordering. If that plan utilizes an index, it might avoid the sort completely which is desirable. But if that plan doesn't use an index it is likely better to filter then sort as we'll have less rows to sort. The observation is that we need to explore both paths.

Reply:
For your example, it seems that the right thing would be for us to have two physical expressions in the memo group for "SELECT * FROM a WHERE > 1" (and only one of them would require an enforcer). The only thing is that we have to be able to represent a requirement for each expr child - the two expressions would be identical except for these requirements. We would get these expressions when applying a logical->physical xform on this group - when we apply the xform we enumerate the possibilities for the children, and each resulting expression can have different physical properties. I am not sure I understood correctly how this should work though.

Reply:
My understanding for this case is that there will be one `select` expression in the group and one `sort` enforcer. Both will be able to satisfy the ordering requirement because the `select` is pass-through on the ordering of the children. Having written that though, I'm having doubts. It is probably useful to have separate `physicalSelect` operators which are nearly identical to `select` except they have physical properties attached. I should probably experiment with this.

Reply:
My understanding is that a *physical* expr represents a plan that has definite physical properties (as opposed to a multitude of possibilities), so I would expect two physical select exprs (and one logical).

Having two operators would allow this, but it seems that we would have to have two versions of most operators. Maybe we should be able to mark each *expr* as physical or not as opposed to having this info in the operator. Only exprs with physical operators can be physical, and exprs with logical operators can't be physical. `select` is both physical and logical so it supports both.

Reply:
That makes sense. And the marking of a physical `select` could be done by specifying a non-nil `expr.physicalProps`. I think `memoExpr.fingerprint()` would just work with this approach as it includes the physical props, if present, in the fingerprint. So for this example we'd have something like:

```
5: [select [1 4]] [select <+0> [1 4]] [select <+1> [1 4]] [sort <+0> [5]]
4: [gt [2 3]]
3: [const 1]
2: [variable a.y]
1: [scan a] [index-scan a@primary <+0> [-]] [inner-join [6 8 -]] [inner-join [8 6 -]]
8: [index-scan a@primary <+0> [7]]
7: [variable a.z]
6: [index-scan a@y_idx <+1> [-]]
```

Reply:
Nice 👍  Also `select <+0> [1 4]` will need to be something like `select <+0> [1<+1> 4]`

Reply:
Did you mean `[select <+0> [1<+0> 4]]`? Select inherits its ordering from its input.

Reply:
Sorry, yes.